common/persistence/cassandra/metadata_store.go

```diff
@@ -161,23 +168,48 @@ func (m *MetadataStore) CreateNamespaceInV2Table(
 
 	previous := make(map[string]interface{})
 	applied, iter, err := m.session.MapExecuteBatchCAS(batch, previous)
+	deleteOrphanNamespace := func() {
+		// Delete namespace from `namespaces_by_id`
+		if errDelete := m.session.Query(templateDeleteNamespaceQuery, request.ID).WithContext(ctx).Exec(); errDelete != nil {
+			m.logger.Warn("Unable to delete orphan namespace record. Error", tag.Error(errDelete))
+		}
+	}
+
 	if err != nil {
 		return nil, serviceerror.NewUnavailable(fmt.Sprintf("CreateNamespace operation failed. Inserting into namespaces table. Error: %v", err))
 	}
+
 	defer func() { _ = iter.Close() }()
 
 	if !applied {
-		// Namespace already exist.  Delete orphan namespace record before returning back to user
-		if errDelete := m.session.Query(templateDeleteNamespaceQuery, request.ID).WithContext(ctx).Exec(); errDelete != nil {
-			m.logger.Warn("Unable to delete orphan namespace record. Error", tag.Error(errDelete))
+
+		// if both conditions fail, find the one related to the first query
+		if name, ok := previous["name"]; !ok || name != request.Name {
+			m := make(map[string]interface{})
+			if iter.MapScan(m) {
+				previous = m
+			}
 		}
 
-		if id, ok := previous["Id"].([]byte); ok {
-			msg := fmt.Sprintf("Namespace already exists.  NamespaceId: %v", primitives.UUIDString(id))
+		// if conditional failure is due to a duplicate name in namespaces table
+		if name, ok := previous["name"]; ok && name == request.Name {
+			existingID := request.ID
+			if id, ok := previous["id"]; ok && gocql.UUIDToString(id) != request.ID {
+				existingID = gocql.UUIDToString(id)
+				// Delete orphan namespace record before returning back to user
+				deleteOrphanNamespace()
+			}
+
+			msg := fmt.Sprintf("Namespace already exists.  NamespaceId: %v", existingID)
 			return nil, serviceerror.NewNamespaceAlreadyExists(msg)
+
 		}
 
-		return nil, serviceerror.NewNamespaceAlreadyExists("CreateNamespace operation failed because of conditional failure.")
+		// If namespace does not exist already and applied is false,
+		// notification_version does not match our expectations and it's conditional failure.
+		// Delete orphan namespace record before returning back to user
+		deleteOrphanNamespace()
```

Comment:
This seems potentially dangerous to me. If we try to create a namespace with the same id twice, and line 170 says that the operation was not applied, but didn't cause an error, we will end up deleting the namespace here.

Reply:
This function is called only from `CreateNamespace` where it checks for uuid collision. It can be problematic if we ever call this directly though. Also, I believe `namespaces` table does not care about `id`, and that makes it hard to deal with same id in this method. 

But now I wonder why we don't add the other query to this batch query 🤔

Reply:
ok since @yycptt mentioned it's not possible to do batch queries across tables, I resolve this conversation.